haskell-src/Concordium/Types/SeedState.hs

```diff
@@ -1,18 +1,58 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
 {-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
 
 module Concordium.Types.SeedState where
 
 import Data.Serialize
-import GHC.Generics
+import Data.Singletons.TH
+import Prelude.Singletons
 
 import Concordium.Crypto.SHA256 (Hash)
 import Concordium.Types
+import Concordium.Types.Conditionally
+
+$( singletons
```

Comment:
Maybe just link to the documentation on `Parameters` for some basic documentation on the singletons and how we intend to use them.